### PR TITLE
Improving prof support on FreeBSD

### PR DESCRIPTION
**Pull request by devnexen**
Dumping process mapping in a native manner.
superseeds #1439

### CODE REVIEW

File: src/prof_data.c [hunk 1]
```diff
@@ -939,13 +944,82 @@ prof_open_maps(const char *format, ...) {
...
+		goto label_return;
+	}
+
+	if (sysctl(mib, 4, NULL, &sz, NULL, 0) == -1) {
+		ret = true;
+		goto label_return;
+	}
+
+	/* Size for kinfo_vmentry */
+	sz = sz * 4 / 3;
```
Comment (davidtgoldblatt) [inline src/prof_data.c]: What's the reasoning behind this computation? Can we have a comment explaining?

File: src/prof_data.c [hunk 2]
```diff
@@ -939,13 +944,82 @@ prof_open_maps(const char *format, ...) {
...
+	buf = mmap(NULL, sz, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANON, -1, 0);
+
+	if (buf == MAP_FAILED) {
+		ret = true;
+		goto label_return;
+	}
+
+	if (sysctl(mib, 4, buf, &sz, NULL, 0) == -1) {
+		ret = true;
+		munmap(buf, sz);
```
Comment (davidtgoldblatt) [inline src/prof_data.c]: Can we unify the munmaps with the mfd cleanup, down at label_return?

File: src/prof_data.c [hunk 3]
```diff
@@ -939,13 +944,82 @@ prof_open_maps(const char *format, ...) {
...
+		goto label_return;
+	}
+
+	char *low = buf;
+	char *high = buf + sz;
+
+	while (low < high) {
+		/* paths + enough room for address ranges and permissions */
+		char map[PATH_MAX+68];
```
Comment (davidtgoldblatt) [inline src/prof_data.c]: Where does 68 come from? Can we add a comment explaining?

File: src/prof_data.c [hunk 4]
```diff
@@ -939,13 +944,82 @@ prof_open_maps(const char *format, ...) {
...
+			ret = true;
+			munmap(buf, sz);
+			goto label_return;
+		}
+
+		if (e->kve_path[0] == '\0') {
+			goto loop_cond;
+		}
+
+		malloc_snprintf(map, sizeof(map), "%.12lu-%.12lu %c%c%c %s\n",
```
Comment (davidtgoldblatt) [inline src/prof_data.c]: Is 12 right? This must be at least nominally enough for the whole address space, right?

File: src/prof_data.c (hunk 1 above)
Comment (devnexen) [reply src/prof_data.c]: Well in fact this is to give enough room for kinfo_vmentry. FreeBSD libpvocstat thus underlyingly nvm does that too.

Comment (davidtgoldblatt) [reply src/prof_data.c]: But like, what property of kinfo_vmentry requires this tweak to the reported size? Is it just to handle races in which a mapping is added in between the two sysctl calls?

File: src/prof_data.c (hunk 4 above)
Comment (devnexen) [reply src/prof_data.c]: Yes it takes not more than 12 large.

File: src/prof_data.c (hunk 1 above)
Comment (devnexen) [reply src/prof_data.c]: All FreeBSD code consumer which list proc maps use this workflow to give enough room for enough process entries in between calls indeed but barely more documented than that.

**Comment by interwq** [comment]
The test failures on Windows look real, e.g. https://ci.appveyor.com/project/jasone/jemalloc/builds/30047453/job/nikr65egvqd1fu3h

File: src/prof_data.c (hunk 4 above)
Comment (davidtgoldblatt) [reply src/prof_data.c]: How come? E.g. if we have a 48-bit address space, it seems we could have as many as ceil(log(2^48)) == 15 digits.

**Comment by davidtgoldblatt** [comment]
I think some variant of this still probably makes sense, but the CI breakages mean probably not this version.

**Pull request closed without merging**